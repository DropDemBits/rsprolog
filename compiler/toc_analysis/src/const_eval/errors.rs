//! Errors during constant evaluation
use toc_span::TextRange;

use crate::const_eval::{ConstValue, RestrictType};

#[derive(Debug, Clone)]
pub struct ConstError {
    kind: ErrorKind,
    span: TextRange,
}

impl ConstError {
    pub(super) fn new(kind: ErrorKind, span: TextRange) -> Self {
        Self { kind, span }
    }

    pub(super) fn without_span(kind: ErrorKind) -> Self {
        Self {
            kind,
            span: TextRange::default(),
        }
    }

    pub(super) fn reported(span: TextRange) -> Self {
        Self {
            kind: ErrorKind::Reported,
            span,
        }
    }

    pub(super) fn change_span(self, new_span: TextRange) -> Self {
        Self {
            kind: self.kind,
            span: new_span,
        }
    }

    pub(super) fn kind(&self) -> &ErrorKind {
        &self.kind
    }

    /// Reports the detailed version of the `ConstError` to the given reporter
    pub fn report_to(&self, reporter: &mut toc_reporting::MessageSink) {
        use toc_reporting::MessageKind;

        // Ignore already reported messages, or for missing expressions
        if matches!(self.kind, ErrorKind::Reported | ErrorKind::MissingExpr) {
            return;
        }

        // Report common message header
        let msg =
            reporter.report_detailed(MessageKind::Error, &format!("{}", self.kind), self.span);

        // Report extra details
        match &self.kind {
            ErrorKind::NoConstExpr(def_span) => {
                // Report at the reference's definition spot
                msg.with_info("reference declared here", *def_span)
            }
            ErrorKind::WrongResultType(found, expected) => {
                let expected_name = match expected {
                    RestrictType::None => panic!("Wrong result type on no restriction"),
                    RestrictType::Integer => "integer value",
                    RestrictType::Real => "real value",
                    RestrictType::Boolean => "boolean value",
                };

                msg.with_note(
                    &format!("expected {}, found {}", expected_name, found.type_name()),
                    self.span,
                )
            }
            _ => msg,
        }
        .finish();
    }
}

#[derive(Debug, Clone, thiserror::Error)]
pub(super) enum ErrorKind {
    // Traversal errors
    /// Encountered an evaluation cycle
    #[error("detected a compile-time evaluation cycle")]
    EvalCycle,
    /// Missing expression operand
    #[error("operand is an invalid expression")]
    MissingExpr,
    /// No const expr is associated with this identifer.
    /// Provided span is the span of the symbol's definition
    #[error("reference cannot be computed at compile-time")]
    NoConstExpr(toc_span::TextRange),
    /// Error is already reported
    #[error("compile-time evaluation error already reported")]
    Reported,

    // Computation errors
    /// Wrong operand type in eval expression
    #[error("wrong type for compile-time expression")]
    WrongOperandType,
    /// Wrong resultant type in eval expression
    #[error("wrong type for compile-time expression")]
    WrongResultType(ConstValue, RestrictType),
    /// Integer overflow
    #[error("integer overflow in compile-time expression")]
    IntOverflow,
    /// Floating point overflow
    #[error("real overflow in compile-time expression")]
    RealOverflow,
    /// Division by zero
    #[error("division by zero in compile-time expression")]
    DivByZero,
    /// Negative int exponent provided during power raising
    #[error("raising integer to a negative exponent")]
    NegativeIntExp,
    /// Negative int shift provided during bit shifting
    #[error("bit shifting integer by a negative amount")]
    NegativeIntShift,

    // Unsupported messages
    /// Currently unsupported const eval value
    #[error("literal is currently not implemented for compile-time evaluation")]
    UnsupportedValue,
    /// Currently unsupported const eval operation
    #[error("operation is currently not implemented for compile-time evaluation")]
    UnsupportedOp,
}
