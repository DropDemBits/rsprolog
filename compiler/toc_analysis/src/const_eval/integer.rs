//! Sign & magnitude integer representation

use std::convert::TryInto;
use std::fmt;

use crate::const_eval::ConstError;

/// Constant Integer representation
#[derive(Debug, Clone, Copy)]
pub struct ConstInt {
    /// invariant: must be representable with the given `sign` and `magnitude`
    magnitude: u64,
    sign: IntSign,
    width: IntWidth,
}

impl ConstInt {
    /// Constructs a new integer constant from an unsigned value
    ///
    /// ## Parameters
    /// - `value`: The unsigned value of the corresponding integer constant
    /// - `allow_64bit_values`: If 64-bit values can be constructed from applying
    ///   operations to this integer constant
    pub fn from_unsigned(value: u64, allow_64bit_ops: bool) -> Result<Self, ConstError> {
        let width = if allow_64bit_ops {
            // Allow 64-bit operations, any value is allowed
            IntWidth::As64
        } else {
            // Apply only 32-bit operations
            if value > u32::MAX as u64 {
                // Already overflowing
                return Err(ConstError::IntOverflow);
            }

            IntWidth::As32
        };

        Ok(Self {
            sign: IntSign::Positive,
            magnitude: value,
            width,
        })
    }

    /// Constructs a new integer constant from a signed floating point value
    ///
    /// ## Parameters
    /// - `value`: The signed floating point value of the corresponding integer constant
    /// - `allow_64bit_values`: If 64-bit values can be constructed from applying
    ///   operations to this integer constant
    pub fn from_signed_real(value: f64, allow_64bit_ops: bool) -> Result<Self, ConstError> {
        let width = if allow_64bit_ops {
            // Allow 64-bit operations, check against u64 bounds
            if (value.is_sign_negative() && value < i64::MIN as f64)
                || (value.is_sign_positive() && value > u64::MAX as f64)
            {
                // Already overflowing
                return Err(ConstError::IntOverflow);
            }

            IntWidth::As64
        } else {
            // Apply only 32-bit operations
            if (value.is_sign_negative() && value < i32::MIN as f64)
                || (value.is_sign_positive() && value > u32::MAX as f64)
            {
                // Already overflowing
                return Err(ConstError::IntOverflow);
            }

            IntWidth::As32
        };

        let (magnitude, sign) = {
            let magnitude = value.abs().trunc() as u64;

            if magnitude == 0 {
                // Always +0
                (0, IntSign::Positive)
            } else {
                let sign = if value.is_sign_negative() {
                    IntSign::Negative
                } else {
                    IntSign::Positive
                };

                (magnitude, sign)
            }
        };

        Ok(Self {
            sign,
            magnitude,
            width,
        })
    }

    /// Converts the `ConstInt` into the corresponding `u32` value.
    ///
    /// ## Returns
    /// Returns `Some(u32)` if the `ConstInt` is a positive integer (including zero)
    /// and is actually representable as a u32, or `None` otherwise.
    pub fn into_u32(self) -> Option<u32> {
        match self.sign {
            IntSign::Positive => self.magnitude.try_into().ok(),
            IntSign::Negative => None,
        }
    }

    /// Converts the `ConstInt` into the corresponding `u64` value.
    ///
    /// ## Returns
    /// Returns `Some(u64)` if the `ConstInt` is a positive integer (including zero),
    /// or `None` otherwise.
    pub fn into_u64(self) -> Option<u64> {
        match self.sign {
            IntSign::Positive => Some(self.magnitude),
            IntSign::Negative => None,
        }
    }

    /// Converts the `ConstInt` into the corresponding `f64` value.
    pub fn into_f64(self) -> f64 {
        match self.sign {
            IntSign::Positive => self.magnitude as f64,
            IntSign::Negative => -(self.magnitude as f64),
        }
    }

    /// Checked integer addition.
    /// Computes `self + rhs`, returning `Err(ConstError::IntOverflow)` if overflow occurred.
    pub fn checked_add(self, rhs: ConstInt) -> Result<ConstInt, ConstError> {
        let effective_width = Self::effective_width(self.width, rhs.width);

        // Potential signs:
        // lhs rhs    magnitude op
        // +   +   => +
        // -   +   => -
        // +   -   => -
        // -   -   => +
        let (value, new_sign) = if self.sign == rhs.sign {
            // Sign is the same as the original
            (self.magnitude.checked_add(rhs.magnitude), self.sign)
        } else {
            // Sign can change
            let (magnitude, sign) = {
                let (magnitude, wrapped) = self.magnitude.overflowing_sub(rhs.magnitude);

                if wrapped {
                    // Undo two's compliment
                    // Flip the sign
                    (!magnitude + 1, self.sign.negate())
                } else {
                    (magnitude, self.sign)
                }
            };

            (Some(magnitude), sign)
        };

        Self::check_overflow(value, new_sign, effective_width)
    }

    /// Checked integer subtraction.
    /// Computes `self - rhs`, returning `Err(ConstError::IntOverflow)` if overflow occurred.
    pub fn checked_sub(self, rhs: ConstInt) -> Result<ConstInt, ConstError> {
        // Reuse addition code
        self.checked_add(rhs.unchecked_negate())
    }

    /// Checked integer multiplication.
    /// Computes `self * rhs`, returning `Err(ConstError::IntOverflow)` if overflow occurred.
    pub fn checked_mul(self, rhs: ConstInt) -> Result<ConstInt, ConstError> {
        let effective_width = Self::effective_width(self.width, rhs.width);

        // Potential signs:
        // lhs rhs    final sign
        // +   +   => +
        // -   +   => -
        // +   -   => -
        // -   -   => +

        // Magnitude is unaffected by the initial signs
        let value = self.magnitude.checked_mul(rhs.magnitude);

        let new_sign = if self.sign == rhs.sign {
            // Will always be positive
            IntSign::Positive
        } else {
            // Will always be negative
            IntSign::Negative
        };

        Self::check_overflow(value, new_sign, effective_width)
    }

    /// Checked integer division.
    /// Computes `self div rhs`, returning `Err(ConstError::DivByZero)` if `rhs == 0`.
    pub fn checked_div(self, rhs: ConstInt) -> Result<ConstInt, ConstError> {
        let effective_width = Self::effective_width(self.width, rhs.width);

        // Potential signs:
        // lhs rhs    final sign
        // +   +   => +
        // -   +   => -
        // +   -   => -
        // -   -   => +

        // Magnitude is unaffected by the initial signs
        let value = self
            .magnitude
            .checked_div(rhs.magnitude)
            .ok_or(ConstError::DivByZero)?;

        let new_sign = if self.sign == rhs.sign {
            // Will always be positive
            IntSign::Positive
        } else {
            // Will always be negative
            IntSign::Negative
        };

        Self::check_overflow(Some(value), new_sign, effective_width)
    }

    /// Negates the sign of the integer.
    /// Does nothing for a magnitude of 0.
    pub fn negate(self) -> Result<ConstInt, ConstError> {
        Self::check_overflow(Some(self.magnitude), self.sign.negate(), self.width)
    }

    /// Negates the sign of the integer.
    /// Does nothing for a magnitude of 0.
    ///
    /// Note: This can break the invariant that for a width of `As32` and sign of `Negative`,
    /// magnitude must be in the range `[0, 0x7FFFFFFF]`.
    /// The only allowed usage of this function is for negating the `rhs` in `checked_sub`
    fn unchecked_negate(mut self) -> Self {
        if self.magnitude > 0 {
            self.sign = self.sign.negate()
        }

        self
    }

    fn effective_width(lhs: IntWidth, rhs: IntWidth) -> IntWidth {
        // Potential widths
        // lhs rhs    effective width
        // 32  32  => 32
        // 64  32  => 32
        // 32  64  => 32
        // 64  64  => 64

        if lhs == IntWidth::As64 && lhs == rhs {
            IntWidth::As64
        } else {
            IntWidth::As32
        }
    }

    fn check_overflow(
        value: Option<u64>,
        new_sign: IntSign,
        effective_width: IntWidth,
    ) -> Result<ConstInt, ConstError> {
        let (magnitude, sign) = {
            let effective_magnitude = value.ok_or(ConstError::IntOverflow)?;

            if effective_magnitude == 0 {
                // `0` is always a "positive" number
                (0, IntSign::Positive)
            } else {
                // Keep the same magnitude
                (effective_magnitude, new_sign)
            }
        };

        // Check for overflow
        let overflowed = match effective_width {
            // [0, 0xFFFF_FFFF]
            IntWidth::As32 if sign.is_positive() => magnitude > u32::MAX as u64,
            // [0, 0x8000_0000]
            IntWidth::As32 if sign.is_negative() => magnitude > i32::MIN.unsigned_abs() as u64,
            // [0, 0xFFFFFFFF_FFFFFFFF] or all values of u64
            IntWidth::As64 if sign.is_positive() => false,
            // [0, 0x80000000_00000000] or all values of u64
            IntWidth::As64 if sign.is_negative() => magnitude > i64::MIN.unsigned_abs() as u64,
            // All cases already covered
            _ => unreachable!(),
        };

        if !overflowed {
            Ok(Self {
                magnitude,
                sign,
                width: effective_width,
            })
        } else {
            Err(ConstError::IntOverflow)
        }
    }
}

impl fmt::Display for ConstInt {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let magnitude = if f.alternate() {
            // Show the width
            let width = match self.width {
                IntWidth::As32 => "i32",
                IntWidth::As64 => "i64",
            };

            format!("{}{}", self.magnitude, width)
        } else {
            // Just the magnitude
            format!("{}", self.magnitude)
        };

        f.pad_integral(matches!(self.sign, IntSign::Positive), "", &magnitude)
    }
}

#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum IntWidth {
    As32,
    As64,
}

#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum IntSign {
    Positive,
    Negative,
}

impl IntSign {
    fn negate(self) -> Self {
        match self {
            IntSign::Positive => IntSign::Negative,
            IntSign::Negative => IntSign::Positive,
        }
    }

    fn is_positive(self) -> bool {
        matches!(self, IntSign::Positive)
    }

    fn is_negative(self) -> bool {
        matches!(self, IntSign::Negative)
    }
}
