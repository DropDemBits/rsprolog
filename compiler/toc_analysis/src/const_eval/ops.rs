//! All valid compile-time operations

use crate::const_eval::{ConstError, ConstInt, ConstValue};

#[derive(Debug, Clone, Copy)]
pub(super) enum ConstOp {
    // Binary operations
    Add,
    Sub,
    Mul,
    Div,
    RealDiv,
    Mod,
    Rem,
    Exp,
    And,
    Or,
    Xor,
    Shl,
    Shr,
    Less,
    LessEq,
    Greater,
    GreaterEq,
    Equal,
    NotEqual,
    Imply,
    // Unary operations
    Not,
    Identity,
    Negate,
}

impl ConstOp {
    pub(super) fn evaluate(
        &self,
        operand_stack: &mut Vec<ConstValue>,
        allow_64bit_ops: bool,
    ) -> Result<ConstValue, ConstError> {
        match self {
            ConstOp::Add => {
                let rhs = operand_stack.pop().unwrap();
                let lhs = operand_stack.pop().unwrap();

                match (lhs, rhs) {
                    (lhs @ ConstValue::Real(_), rhs) | (lhs, rhs @ ConstValue::Real(_)) => {
                        let (lhs, rhs) = (lhs.cast_into_real()?, rhs.cast_into_real()?);

                        match lhs + rhs {
                            v if v.is_infinite() => Err(ConstError::RealOverflow),
                            v => Ok(ConstValue::Real(v)),
                        }
                    }
                    (lhs @ ConstValue::Integer(_), rhs) | (lhs, rhs @ ConstValue::Integer(_)) => {
                        let (lhs, rhs) = (lhs.cast_into_int()?, rhs.cast_into_int()?);
                        lhs.checked_add(rhs).map(|v| ConstValue::Integer(v))
                    }
                    _ => Err(ConstError::WrongType),
                }
            }
            ConstOp::Sub => {
                let rhs = operand_stack.pop().unwrap();
                let lhs = operand_stack.pop().unwrap();

                match (lhs, rhs) {
                    (lhs @ ConstValue::Real(_), rhs) | (lhs, rhs @ ConstValue::Real(_)) => {
                        let (lhs, rhs) = (lhs.cast_into_real()?, rhs.cast_into_real()?);

                        match lhs - rhs {
                            v if v.is_infinite() => Err(ConstError::RealOverflow),
                            v => Ok(ConstValue::Real(v)),
                        }
                    }
                    (lhs @ ConstValue::Integer(_), rhs) | (lhs, rhs @ ConstValue::Integer(_)) => {
                        let (lhs, rhs) = (lhs.cast_into_int()?, rhs.cast_into_int()?);
                        lhs.checked_sub(rhs).map(|v| ConstValue::Integer(v))
                    }
                    _ => Err(ConstError::WrongType),
                }
            }
            ConstOp::Mul => {
                let rhs = operand_stack.pop().unwrap();
                let lhs = operand_stack.pop().unwrap();

                match (lhs, rhs) {
                    (lhs @ ConstValue::Real(_), rhs) | (lhs, rhs @ ConstValue::Real(_)) => {
                        let (lhs, rhs) = (lhs.cast_into_real()?, rhs.cast_into_real()?);

                        match lhs * rhs {
                            v if v.is_infinite() => Err(ConstError::RealOverflow),
                            v => Ok(ConstValue::Real(v)),
                        }
                    }
                    (lhs @ ConstValue::Integer(_), rhs) | (lhs, rhs @ ConstValue::Integer(_)) => {
                        let (lhs, rhs) = (lhs.cast_into_int()?, rhs.cast_into_int()?);
                        lhs.checked_mul(rhs).map(|v| ConstValue::Integer(v))
                    }
                    _ => Err(ConstError::WrongType),
                }
            }
            ConstOp::Div => {
                let rhs = operand_stack.pop().unwrap();
                let lhs = operand_stack.pop().unwrap();

                match (lhs, rhs) {
                    (lhs @ ConstValue::Real(_), rhs) | (lhs, rhs @ ConstValue::Real(_)) => {
                        let (lhs, rhs) = (lhs.cast_into_real()?, rhs.cast_into_real()?);

                        // Divide & truncate, then convert to an integer
                        match (lhs / rhs).trunc() {
                            _ if rhs == 0.0 => Err(ConstError::DivByZero),
                            v => ConstInt::from_signed_real(v, allow_64bit_ops)
                                .map(|v| ConstValue::Integer(v)),
                        }
                    }
                    (lhs @ ConstValue::Integer(_), rhs) | (lhs, rhs @ ConstValue::Integer(_)) => {
                        let (lhs, rhs) = (lhs.cast_into_int()?, rhs.cast_into_int()?);
                        lhs.checked_div(rhs).map(|v| ConstValue::Integer(v))
                    }
                    _ => Err(ConstError::WrongType),
                }
            }
            ConstOp::RealDiv => todo!(),
            ConstOp::Mod => todo!(),
            ConstOp::Rem => todo!(),
            ConstOp::Exp => todo!(),
            ConstOp::And => {
                let rhs = operand_stack.pop().unwrap();
                let lhs = operand_stack.pop().unwrap();

                match (lhs, rhs) {
                    (lhs @ ConstValue::Integer(_), rhs) | (lhs, rhs @ ConstValue::Integer(_)) => {
                        let (lhs, rhs) = (lhs.cast_into_int()?, rhs.cast_into_int()?);
                        Ok(ConstValue::Integer(lhs.and(rhs)))
                    }
                    (lhs @ ConstValue::Bool(_), rhs) | (lhs, rhs @ ConstValue::Bool(_)) => {
                        let (lhs, rhs) = (lhs.cast_into_bool()?, rhs.cast_into_bool()?);
                        Ok(ConstValue::Bool(lhs & rhs))
                    }
                    _ => Err(ConstError::WrongType),
                }
            }
            ConstOp::Or => {
                let rhs = operand_stack.pop().unwrap();
                let lhs = operand_stack.pop().unwrap();

                match (lhs, rhs) {
                    (lhs @ ConstValue::Integer(_), rhs) | (lhs, rhs @ ConstValue::Integer(_)) => {
                        let (lhs, rhs) = (lhs.cast_into_int()?, rhs.cast_into_int()?);
                        Ok(ConstValue::Integer(lhs.or(rhs)))
                    }
                    (lhs @ ConstValue::Bool(_), rhs) | (lhs, rhs @ ConstValue::Bool(_)) => {
                        let (lhs, rhs) = (lhs.cast_into_bool()?, rhs.cast_into_bool()?);
                        Ok(ConstValue::Bool(lhs | rhs))
                    }
                    _ => Err(ConstError::WrongType),
                }
            }
            ConstOp::Xor => {
                let rhs = operand_stack.pop().unwrap();
                let lhs = operand_stack.pop().unwrap();

                match (lhs, rhs) {
                    (lhs @ ConstValue::Integer(_), rhs) | (lhs, rhs @ ConstValue::Integer(_)) => {
                        let (lhs, rhs) = (lhs.cast_into_int()?, rhs.cast_into_int()?);
                        Ok(ConstValue::Integer(lhs.xor(rhs)))
                    }
                    (lhs @ ConstValue::Bool(_), rhs) | (lhs, rhs @ ConstValue::Bool(_)) => {
                        let (lhs, rhs) = (lhs.cast_into_bool()?, rhs.cast_into_bool()?);
                        Ok(ConstValue::Bool(lhs ^ rhs))
                    }
                    _ => Err(ConstError::WrongType),
                }
            }
            ConstOp::Shl => todo!(),
            ConstOp::Shr => todo!(),
            ConstOp::Less => todo!(),
            ConstOp::LessEq => todo!(),
            ConstOp::Greater => todo!(),
            ConstOp::GreaterEq => todo!(),
            ConstOp::Equal => todo!(),
            ConstOp::NotEqual => todo!(),
            ConstOp::Imply => {
                let rhs = operand_stack.pop().unwrap();
                let lhs = operand_stack.pop().unwrap();

                match (lhs, rhs) {
                    (lhs @ ConstValue::Bool(_), rhs) | (lhs, rhs @ ConstValue::Bool(_)) => {
                        let (lhs, rhs) = (lhs.cast_into_bool()?, rhs.cast_into_bool()?);
                        Ok(ConstValue::Bool(!lhs | rhs))
                    }
                    _ => Err(ConstError::WrongType),
                }
            }
            ConstOp::Not => {
                let rhs = operand_stack.pop().unwrap();

                match rhs {
                    ConstValue::Integer(v) => Ok(ConstValue::Integer(v.not())),
                    ConstValue::Bool(v) => Ok(ConstValue::Bool(!v)),
                    _ => Err(ConstError::WrongType),
                }
            }
            ConstOp::Identity => {
                // Rhs must be a number
                let rhs = operand_stack.pop().unwrap();

                match rhs {
                    rhs @ ConstValue::Integer(_) => Ok(rhs),
                    rhs @ ConstValue::Real(_) => Ok(rhs),
                    _ => Err(ConstError::WrongType),
                }
            }
            ConstOp::Negate => {
                let rhs = operand_stack.pop().unwrap();

                match rhs {
                    ConstValue::Integer(v) => v.negate().map(|v| ConstValue::Integer(v)),
                    ConstValue::Real(v) => Ok(ConstValue::Real(-v)),
                    _ => Err(ConstError::WrongType),
                }
            }
        }
    }
}
