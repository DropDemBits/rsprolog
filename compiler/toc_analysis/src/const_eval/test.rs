use std::sync::Arc;

use unindent::unindent;

use crate::const_eval::ConstEvalCtx;

#[track_caller]
fn assert_const_eval(source: &str) {
    insta::assert_snapshot!(insta::internals::AutoName, do_const_eval(source), source);
}

/// Expr version
#[track_caller]
fn assert_const_eval_expr(expr: &str) {
    let source = format!("const _ := {}", expr);
    insta::assert_snapshot!(insta::internals::AutoName, do_const_eval(&source), &expr);
}

macro_rules! for_all_const_exprs {
    ($($src:literal)+) => {
        $(assert_const_eval_expr($src);)+
    };
}

fn do_const_eval(source: &str) -> String {
    let parsed = toc_parser::parse(&source);
    let mut unit_map = toc_hir::UnitMapBuilder::new();
    let hir_res = toc_hir_lowering::lower_ast(parsed.syntax(), &mut unit_map);
    let unit_map = Arc::new(unit_map.finish());

    let unit = unit_map.get_unit(hir_res.id);
    let const_eval_ctx = Arc::new(ConstEvalCtx::new(unit_map.clone()));
    super::collect_const_vars(unit, const_eval_ctx.clone());

    let mut results_str = String::new();

    // Need access to the inner state of the ConstEvalCtx, which is behind a lock
    {
        // Eagerly evaluate all of the available const vars
        let mut inner = const_eval_ctx.inner.write().unwrap();
        let const_exprs = inner.var_to_expr.values().copied().collect::<Vec<_>>();

        for expr in const_exprs {
            let results = match inner.eval_expr(expr) {
                Ok(v) => format!("{:?} -> {:?}\n", expr, v),
                Err(err) => format!("{:?} -> {:?}\n", expr, err),
            };

            results_str.push_str(&results);
        }
    }

    // Errors are bundled into the const error context
    stringify_const_eval_results(&results_str, &const_eval_ctx)
}

fn stringify_const_eval_results(results: &str, const_eval: &ConstEvalCtx) -> String {
    // Pretty print const eval ctx
    format!("{:#?}\n{}", const_eval, results)
}

#[test]
fn complex_arithmetic_expr() {
    assert_const_eval_expr("1 + 2 * 3 div 3 - 4 + 5");
}

#[test]
fn arithmetic_const_ops() {
    // All operations should be evaluated
    for_all_const_exprs![
        "1 + 1"
        "10 - 2"
        "2 * 5"
        "3 div 2"
        "-1"
        "+-1"

        "1 + 1.0"
        "10 - 2.0"
        "2 * 5.0"
        "3 div 2.0"
        "-1.0"
        "+-1.0"
    ];
}

#[test]
fn logic_const_ops() {
    for_all_const_exprs![
        "false and false" // 0
        "false and true"  // 0
        "true and false"  // 0
        "true and true"   // 1

        "false or false"  // 0
        "false or true"   // 1
        "true or false"   // 1
        "true or true"    // 1

        "false xor false" // 0
        "false xor true"  // 1
        "true xor false"  // 1
        "true xor true"   // 0

        "false => false"  // 1
        "false => true"   // 1
        "true => false"   // 0
        "true => true"    // 1

        "not true"
        "not false"
    ];
}

#[test]
fn bitwise_const_ops() {
    for_all_const_exprs![
        "0 and 0"
        "0 and 1"
        "1 and 0"
        "1 and 1"

        "0 or 0"
        "0 or 1"
        "1 or 0"
        "1 or 1"

        "0 xor 0"
        "0 xor 1"
        "1 xor 0"
        "1 xor 1"

        "not 0"
        "not 1"

        // Applicable to negative numbers too
        "0 and -0"
        "0 and -1"
        "1 and -0"
        "1 and -1"

        "0 or -0"
        "0 or -1"
        "1 or -0"
        "1 or -1"

        "0 xor -0"
        "0 xor -1"
        "1 xor -0"
        "1 xor -1"

        "not -0"
        "not -1"
    ];
}

#[test]
fn real_promotion() {
    for_all_const_exprs![
        "1.0 + 1"
        "  1 + 1.0"
        "1.0 + 1.0"

        "1.0 - 1"
        "  1 - 1.0"
        "1.0 - 1.0"

        "1.0 * 1"
        "  1 * 1.0"
        "1.0 * 1.0"

        "1.0 div 1"
        "  1 div 1.0"
        "1.0 div 1.0"
    ];
}

#[test]
fn const_local_var_lookup() {
    assert_const_eval_expr(&unindent(
        r#"
    const a := 1
    const b := a + 1
    const c := b
    "#,
    ));
}

#[test]
fn error_div_by_zero() {
    // TODO: Add tests cases for real div, mod & rem
    for_all_const_exprs!["1 div 0"];
    for_all_const_exprs!["1 div 0.0"];
    for_all_const_exprs!["1.0 div 0"];
    for_all_const_exprs!["1.0 div 0.0"];
}

#[test]
fn error_int_overflow_32bit() {
    // TODO: Test operations on 64-bit values

    for_all_const_exprs![
        // 64-bit literals should overflow right away
        "16#100000000"
        // Unsigned 32-bit literals shouldn't
        "16#80000000 % no overflow"
        "16#FFFFFFFF % no overflow"
        // Over add
        "16#ffffffff + 1"
        "-16#80000000 + -1"
        // Over sub
        "16#ffffffff - (-1)"
        "-16#80000000 - 1"
        "16#ffffffff - 16#ffffffff % no overflow"
        // Over mul (positive result)
        "16#80000000 * 2"
        "-16#80000000 * -1 % no overflow"
        // Over mul (negative result)
        "16#80000001 * -1"
        "-16#80000000 * 2"
        // Over idiv (positive result)
        "1e100 div 1"
        "1e100 div 1.0"
        // Over idiv (negative result)
        "16#FFFFFFFF div -1"
        "16#80000001 div -1"
        // Over negate
        "-16#FFFFFFFF"
        "--16#80000000 % no overflow"
        // Over identity (no overflow)
        "+16#FFFFFFFF % no overflow"
        "+-16#80000000 % no overflow"
    ];
}

#[test]
fn error_real_overflow() {
    for_all_const_exprs![
        // Over add
        "1e308 + 1e308"
        "-1e308 + (-1e308)"
        // Over sub
        "1e308 - (-1e308)"
        "-1e308 - 1e308"
        // Over mul
        "1e308 * 10"
        "-1e308 * 10"
        // Over idiv (checked in `error_int_overflow_{32,64}_bit`)
    ];
}

#[test]
fn error_arithmetic_wrong_types() {
    for_all_const_exprs![
        "1 + true"
        "1 - true"
        "1 * true"
        "1 div true"
        "false + 1"
        "false - 1"
        "false * 1"
        "false div 1"

        "1.0 + true"
        "1.0 - true"
        "1.0 * true"
        "1.0 div true"
        "false + 1.0"
        "false - 1.0"
        "false * 1.0"
        "false div 1.0"

        "-false"
        "+true"
    ];
}

#[test]
fn error_logical_wrong_types() {
    for_all_const_exprs![
        "1.0 and false"
        "1.0 or false"
        "1.0 xor false"
        "1.0 => false"

        "false and 1.0"
        "false or 1.0"
        "false xor 1.0"
        "false => 1.0"

        "not 1.0"
    ];
}

#[test]
fn error_no_const_expr() {
    // Referencing a runtime-evaluated var
    assert_const_eval(&unindent(
        r#"
    var a := 1
    const b := a
    "#,
    ));

    // Referencing a runtime-evaluated const
    assert_const_eval(&unindent(
        r#"
    var a := 1
    const b := a
    const c := b
    "#,
    ));

    // Referencing `self`
    assert_const_eval(&unindent(
        r#"
    const a := self
    "#,
    ));
}

#[test]
fn error_propogation() {
    // Propogation of errors
    assert_const_eval(&unindent(
        r#"
    const a := 1 div 0
    const b := a
    "#,
    ));
}
