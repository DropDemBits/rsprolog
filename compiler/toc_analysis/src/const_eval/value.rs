//! Compile-time values

use crate::const_eval::{errors::ErrorKind, ConstError, ConstInt, RestrictType};

#[derive(Debug, Clone)]
pub enum ConstValue {
    /// General integer value
    Integer(ConstInt),
    /// Floating point value
    Real(f64),
    /// Boolean value
    Bool(bool),
}

impl ConstValue {
    /// Unwraps a `ConstValue` into the corresponding `ConstInt`
    ///
    /// ## Returns
    /// If `self` is a `ConstValue::Integer`, returns the corresponding ConstInt value.
    /// Otherwise, returns `ConstError::WrongType`.
    pub fn into_int(self) -> Result<ConstInt, ConstError> {
        match self {
            ConstValue::Integer(v) => Ok(v),
            _ => Err(ConstError::new(ErrorKind::WrongResultType(
                self,
                RestrictType::Integer,
            ))),
        }
    }

    /// Gets the human readable version of the value's type
    pub fn type_name(&self) -> &str {
        match self {
            ConstValue::Integer(_) => "integer value",
            ConstValue::Real(_) => "real value",
            ConstValue::Bool(_) => "boolean value",
        }
    }

    /// Converts a `ConstValue` into a `ConstInt`.
    ///
    /// The only value types that are allowed to be cast into `ConstInt` are:
    ///
    /// - `Integer`
    pub(super) fn cast_into_int(self) -> Result<ConstInt, ConstError> {
        match self {
            ConstValue::Integer(v) => Ok(v),
            _ => Err(ConstError::new(ErrorKind::WrongOperandType)),
        }
    }

    /// Converts a `ConstValue` into a `f64`.
    ///
    /// The only value types that are allowed to be cast into a `f64` are:
    ///
    /// - `Integer`
    /// - `Real`
    pub(super) fn cast_into_real(self) -> Result<f64, ConstError> {
        match self {
            ConstValue::Integer(v) => Ok(v.into_f64()),
            ConstValue::Real(v) => Ok(v),
            _ => Err(ConstError::new(ErrorKind::WrongOperandType)),
        }
    }

    /// Converts a `ConstValue` into a `bool`.
    ///
    /// The only value types that are allowed to be cast into `bool` are:
    ///
    /// - `Bool`
    pub(super) fn cast_into_bool(self) -> Result<bool, ConstError> {
        match self {
            ConstValue::Bool(v) => Ok(v),
            _ => Err(ConstError::new(ErrorKind::WrongOperandType)),
        }
    }
}
