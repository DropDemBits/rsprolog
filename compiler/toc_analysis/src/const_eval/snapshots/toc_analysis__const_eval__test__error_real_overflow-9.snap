---
source: compiler/toc_analysis/src/const_eval/test.rs
expression: 10.0 ** 309

---
ConstEvalCtx {
    inner: RwLock {
        data: InnerCtx {
            eval_infos: [
                EvalInfo { unit_id: UnitId(0), span: 11..22, state: Error(ConstError { kind: RealOverflow, span: 16..18 }) },
            ],
            var_to_expr: {
                GlobalDefId(UnitId(0), DefId(0)): ConstExpr { id: 0 },
            },
        },
    },
}
ConstExpr { id: 0 } -> ConstError { kind: RealOverflow, span: 16..18 }

error at 16..18: real overflow in compile-time expression

